### ISSUE TITLE
Jetpack compose

### ISSUE THREAD

**Issue by OxygenCobalt** [OxygenCobalt/Auxio#901]
Give up and move literally everything but the nightmarish bottom sheet scaffold to Jetpack Compose.

I'm just tired of the state management issues in the current UI. Working with React during my summer internship and on other projects has been actually enjoyable, compared to the current View model. It would also allow me to do nice little animations easier and actually keep up with Android Studio's rapid deprecation of XML.

I'll try to do it within reason, maintaining the same exact user flow fundamentally and trying to avoid making the app feel "weird". Would guess this co-incides with #761.

**Comment by realjoni17** [comment]
Hi i wanna contribute to it

**Comment by OxygenCobalt** [comment]
Cool @realjoni17, although I don't think the app is in a good state for this yet. Compose is somewhat in flux and has some weird UX differences that I would need to hack around to maintain parity with the existing app, so it would be a very delicate and gradual migration.

Additionally, I would also need to do some drastic refactors to the current app architecture just so Compose's idioms actually work, hence #761. Will let you know when the app is good to go for this.

**Comment by Koitharu** [comment]
I hope this app will not migrate to compose

**Comment by OxygenCobalt** [comment]
Hey @Koitharu, do you have anything to share on compose's drawbacks? The most I'm aware of is that it has similar problems to i.e flutter apps, where some components just don't feel quite right. But I think like basic lists and detail views would be able to benefit from compose and they seem pretty native from what I know.

**Comment by navneet851** [comment]
but compose have great upperhand upon xml . due to its flexibility and easy to maintain.

i also wanted to contribute let me know if its possible.

**Comment by OxygenCobalt** [comment]
I'm just...hesistant @navneet851. Lots of things in android "look" simple and ergonomic and then crumple into something worse than the "old manual system" once you apply it to something real-world and complex. I have a really high bar for UI stability and if compose doesn't meet it I'm better off with normal XML.

**Comment by navneet851** [comment]
once you try it you would never look back , i guarantee it. its that simple and effective.

**Comment by navneet851** [comment]
you can watch my song streaming app UI on my github all made upon jetpack compose.

**Comment by zalnaRs** [comment]
> Hey @Koitharu, do you have anything to share on compose's drawbacks? The most I'm aware of is that it has similar problems to i.e flutter apps, where some components just don't feel quite right. But I think like basic lists and detail views would be able to benefit from compose and they seem pretty native from what I know.

It's like flutter. It's a game engine, and it lags like I have the first android phone ever released.

**Comment by OxygenCobalt** [comment]
> > Hey @Koitharu, do you have anything to share on compose's drawbacks? The most I'm aware of is that it has similar problems to i.e flutter apps, where some components just don't feel quite right. But I think like basic lists and detail views would be able to benefit from compose and they seem pretty native from what I know.
> 
> It's like flutter. It's a game engine, and it lags like I have the first android phone ever released. The DX isn't that good either, but Auxio would still be better with compose though as most Android music players are crap.

Yikes, and here I thought it compiled down into Android XML @zalnaRs. Considering my alternative plan to reduce XML spaghetti code then.

**Comment by OxygenCobalt** [comment]
No wonder I've noticed compose apps have the same "someone replaced all your furniture with almost identical copies with slight differences" problem I've noticed with flutter.

**Comment by OxygenCobalt** [comment]
I don't think this is going to happen unless I can confirm that there is total UX parity (design and efficiency) between XML and Compose. Anything that degrades experience will make this a no-go, and I'll continue to maintain the XML.

**Comment by OxygenCobalt** [comment]
@zalnaRs, if you're willing, I might prototype a compose layout and bring it past you to see if it's less efficient than XML. Apparently performance is more a product of state reads and recomposition more than graphics it seems.

Frankly, I think even the mid DX of compose beats out the fragile and hacked together bottom sheet spaghetti code I have to maintain.

**Comment by zalnaRs** [comment]
@OxygenCobalt if XML maintenance becomes hard just rewrite it to compose. It isn't worth losing this project, and as I mentioned earlier Auxio will always be the best.

**Comment by OxygenCobalt** [comment]
> the issue is the jankiness of animations and components not looking right

*Ah.* That's what I noticed too. Will see what can be done.

**Comment by OxygenCobalt** [comment]
> the issue is the jankiness of animations and components not looking right

*Ah.* That's what I noticed too @zalnaRs. Will see what can be done. There are likely obsessive workarounds I can do to maintain parity that will now at least neatly wrap up into abstractions, rather than leak all over the codebase.

**Comment by OxygenCobalt** [comment]
Do you actually know what components look kinda off @zalnaRs? This might allow me to target those during my experiments and ID why they're weird (I've already had to spelunk androidx source code before to fix UI issues, this is nothing new).

**Comment by zalnaRs** [comment]
> Do you actually know what components look kinda off @zalnaRs? This might allow me to target those during my experiments and ID why they're weird (I've already had to spelunk androidx source code before to fix UI issues, this is nothing new).

For example the button's ripple animation, the scaffold (appbar etc...)

**Comment by KraXen72** [comment]
You could also just move *some* parts of the ui into compose, like the bottom sheet. This type of incremental migration to compose is what the Tachiyomi (now Mihon) app did and it worked okay (as far as the iterop goes) iirc

**Comment by OxygenCobalt** [comment]
For now I'll from-scratching the compose UI in my off time @KraXen72, but that seems like a good idea.

So far, I've managed to do a good bit of the bottom sheet system with basically no brittle spaghetti code and a layout system that is actually okay with what I'm doing. It's a little off-standard but I'll figure out the problems. I think having a hybrid approach will work well, at least until I can figure out the discrepancies in the scaffold/button designs.

**Comment by KraXen72** [comment]
good luck! and i hope the 4.0.0 release goes as planned (and soon :))

**Comment by OxygenCobalt** [comment]
MDC-android is in maintenance mode.

<img width="965" height="704" alt="Image" src="https://github.com/user-attachments/assets/d827387e-9448-460d-9bb7-2e50971e7d2a" />

I guess I have to do this now.

**Comment by KraXen72** [comment]
Good Luck! FWIW compose has met the hardware half way and I no longer have any lag on compose-based apps on my 2-3 year old phone, as i used to have a couple of years ago on my old phone. 
I'd even say, since this is maintainance mode (and not unmaintained), considering your life updates in that one issue, take your time, man. Finish all the features/refactors/cleanups you wanted to make in this upcoming release, and take a well-deserved break :)
